hive/src/test/java/org/apache/iceberg/hive/HiveCreateReplaceTableTest.java

```diff
@@ -0,0 +1,211 @@
+    metastore.start();
+    HiveCreateReplaceTableTest.hiveConf = metastore.hiveConf();
+    HiveCreateReplaceTableTest.metastoreClient = new HiveMetaStoreClient(hiveConf);
+    String dbPath = metastore.getDatabasePath(DB_NAME);
+    Database db = new Database(DB_NAME, "description", dbPath, new HashMap<>());
+    metastoreClient.createDatabase(db);
+    HiveCreateReplaceTableTest.catalog = new HiveCatalog(hiveConf);
+  }
+
+  @AfterClass
+  public static void stopMetastore() {
+    catalog.close();
+    HiveCreateReplaceTableTest.catalog = null;
+
+    metastoreClient.close();
+    HiveCreateReplaceTableTest.metastoreClient = null;
+
+    metastore.stop();
+    HiveCreateReplaceTableTest.metastore = null;
+  }
+
+  @Before
+  public void createTableLocation() throws IOException {
+    tableLocation = temp.newFolder("hive-").getPath();
+  }
+
+  @After
+  public void cleanup() {
+    catalog.dropTable(TABLE_IDENTIFIER);
+  }
+
+  @Test
+  public void testCreateTableTxn() {
+    Assert.assertFalse("Table should not exist", catalog.tableExists(TABLE_IDENTIFIER));
+
+    Transaction txn = catalog.newCreateTableTransaction(
+        TABLE_IDENTIFIER, SCHEMA, SPEC, tableLocation, Maps.newHashMap());
+    txn.updateProperties()
+        .set("prop", "value")
+        .commit();
+
+    // verify the table is still not visible before the transaction is committed
+    Assert.assertFalse(catalog.tableExists(TABLE_IDENTIFIER));
+
+    txn.commitTransaction();
+
+    Table table = catalog.loadTable(TABLE_IDENTIFIER);
+    Assert.assertEquals("Table props should match", "value", table.properties().get("prop"));
+  }
+
+  @Test
+  public void testCreateTableTxnTableCreatedConcurrently() {
+    exceptionRule.expect(RuntimeException.class);
+    exceptionRule.expectMessage("Metastore operation failed");
+
+    Assert.assertFalse("Table should not exist", catalog.tableExists(TABLE_IDENTIFIER));
+
+    Transaction txn = catalog.newCreateTableTransaction(
+        TABLE_IDENTIFIER, SCHEMA, SPEC, tableLocation, Maps.newHashMap());
+
+    // create the table concurrently
+    catalog.createTable(TABLE_IDENTIFIER, SCHEMA, SPEC);
+    Assert.assertTrue("Table should be created", catalog.tableExists(TABLE_IDENTIFIER));
+
+    // expect the transaction to fail
+    txn.commitTransaction();
+  }
+
+  @Test
+  public void testCreateTableTxnTableAlreadyExists() {
+    exceptionRule.expect(AlreadyExistsException.class);
+    exceptionRule.expectMessage("Table already exists: hivedb.tbl");
+
+    Assert.assertFalse("Table should not exist", catalog.tableExists(TABLE_IDENTIFIER));
+
+    // create a table before starting a transaction
+    catalog.createTable(TABLE_IDENTIFIER, SCHEMA, SPEC);
+    Assert.assertTrue("Table should be created", catalog.tableExists(TABLE_IDENTIFIER));
+
+    Transaction txn = catalog.newCreateTableTransaction(
+        TABLE_IDENTIFIER, SCHEMA, SPEC, tableLocation, Maps.newHashMap());
```

Comment:
Should the exception be thrown in `newCreateTableTransaction` instead?

Reply:
It is already, I forgot to update the test.